h2/src/main/org/h2/util/ParserUtil.java

```diff
@@ -93,12 +93,13 @@ public static int getSaveTokenType(String s, boolean functionsAsKeywords) {
         case 'A':
             return getKeywordOrIdentifier(s, "ALL", KEYWORD);
         case 'C':
-            if ("CHECK".equals(s)) {
-                return KEYWORD;
-            } else if ("CONSTRAINT".equals(s)) {
-                return KEYWORD;
-            } else if ("CROSS".equals(s)) {
-                return KEYWORD;
+            switch (s) {
+                case "CHECK":
+                    return KEYWORD;
+                case "CONSTRAINT":
+                    return KEYWORD;
+                case "CROSS":
+                    return KEYWORD;
```

Comment:
Actulally `switch` may be slower here, it's better to run JMH tests before such changes.

Reply:
do you have any JMH performance test cases for H2 I could use to check for potential regressions I should I prepare some myself?

If not I'll prepare some JMH for this method.

Reply:
JMH is not for testing overall performance of large projects. It's for testing small methods. But to perform such test for such pieces of code you should ensure that you use a brand new string on each execution. Constant string are handled in faster way in `switch` because they already knows their hash code and result with them will be incorrect.

I run some tests with this specific part with `CONSTRAINT` and `CROSS` and `switch` is slower than a pair of `equals()` by a 30% on Java 9. This was expected. `Parser` is a performance-critical part of H2, so this slowdown is not good.

Reply:
I prepared a JMH microbenchmark: https://github.com/jlawrynowicz/jmh-switch-microbenchmark

You were actually spot on - both on JDK7 and JDK9 version with first switch on character and then if/else was actually slightly faster then current version, one I suggested here and even simplified one (only one switch) and pure if/else implementation.

Weirdly enough, I also tested using static hashMap which was actually much faster (~49% for openjdk7, ~45% on oracle jdk9). 

I'll revert all the switch changes as they don't seem to be such a perf win as I thought they were, but perhaps we could look into the hashMap numbers - perhaps somethings too unrealistic with my benchmark.

```
# tested on linux 4.15.9, CPU: i7-4710HQ

# OpenJDK Runtime Environment (IcedTea 2.6.13) (Arch Linux build 7.u171_2.6.13-1-x86_64)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  58.551 ± 1.023  ops/s
JmhMicrobenchmark.if_else              thrpt   10  36.468 ± 0.313  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  38.419 ± 0.687  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  37.635 ± 0.699  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  39.302 ± 0.728  ops/s

# Java(TM) SE Runtime Environment (build 9.0.4+11)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  56.675 ± 0.400  ops/s
JmhMicrobenchmark.if_else              thrpt   10  36.036 ± 0.620  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  37.570 ± 0.613  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  35.304 ± 0.601  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  39.118 ± 0.566  ops/s
```

Reply:
It's not easy to test hash-based implementations (including switch with strings) correctly.

`String.hashCode()` may be slow enough on the first invocation. There is a big difference between newly allocated strings and old string that already has computed hash code. But `ParserUtil` may got a new string without a hash code and may got a string from cache with known hash code and proportions of such variants are unknown.

You should pass at least some strings that are allocated with `new String(oldString.toCharArray())` or similar methods. `new String(oldString)` copies hash code from the old string and should not be used due to this reason. But allocation of such new strings should be excluded from the testing times. So proper test will be complex enough.

Also different input values are needed including custom strings that are not keywords.

From my point of view there is no sense to spend more time for such tests. It will take a lot of time and looks like there is no way to achieve a really significant performance win.

Reply:
I reran the benchmark with worst case scenario of all strings newly created (should be even as all tests have the same number of new String(s.toCharArray()) calls and the results are again in favor of current implementation. It goes to show how well optimized it already was.

I guess without realistic, synthetic benchmark it's hard to think about optimizing it. I'd believe that due to java's string cache most strings would already have their hash code precalculated, so HashMap approach could be faster, but I have no hard data to back it up at the moment. I'll stop looking into it.

I already crunched the numbers, so I might as well post them for someone perhaps:
```
# tested on linux 4.15.9, CPU: i7-4710HQ

# OpenJDK Runtime Environment (IcedTea 2.6.13) (Arch Linux build 7.u171_2.6.13-1-x86_64)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  17.648 ± 0.536  ops/s
JmhMicrobenchmark.if_else              thrpt   10  20.526 ± 0.169  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  20.972 ± 0.167  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  19.414 ± 0.191  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  21.951 ± 0.192  ops/s

# Java(TM) SE Runtime Environment (build 9.0.4+11)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  17.202 ± 0.210  ops/s
JmhMicrobenchmark.if_else              thrpt   10  20.604 ± 0.148  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  19.990 ± 0.115  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  18.297 ± 0.086  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  20.790 ± 0.248  ops/s
```